File: 142_detectCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {

    public ListNode detectCycle(ListNode head) {
    
	    if (head == null || head.next == null) {
			return null;
		}

		ListNode slow = head;
		ListNode fast = head;

		while (fast != null && fast.next != null) {

			slow = slow.next;
			fast = fast.next.next;

			if (slow == fast) {
				ListNode head2 = head;
				while (head2 != slow) {
					slow = slow.next;
					head2 = head2.next;
				}
				return slow;
			}
		}

		return null;
    }
}

File: 142_detectCycle.go
/
*
 * @lc app=leetcode.cn id=142 lang=golang
 *
 * [142] 环形链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {

  if head == nil {
    return head
  }

  dummy := new(ListNode)
  dummy.Next = head
  slow := dummy
  fast := dummy

  for slow.Next != nil && fast.Next != nil && fast.Next.Next !=nil {

    slow = slow.Next
    fast = fast.Next.Next

    if slow == fast {

      fast = dummy
      for fast != slow {
        fast = fast.Next
        slow = slow.Next
      }
      return slow
    }
  }
  
  return nil
}
// @lc code=end

